src/daft-core/src/series/from_lit.rs

```diff
@@ -53,6 +93,422 @@ pub(crate) fn combine_lit_types(left: &DataType, right: &DataType) -> Option<Dat
     }
 }
 
+/// Creates a series from an iterator of `Result<Literal>`.
+/// It also captures any errors occurred during iteration
+/// It returns a Series, along with an IndexMap of errors
+// TODO(cory): combine this and `from_literals`
+pub fn series_from_literals_iter<I: ExactSizeIterator<Item = DaftResult<Literal>> + TrustedLen>(
+    values: I,
+    dtype: DataType,
+) -> DaftResult<(Series, Option<IndexMap<usize, String>>)> {
+    let downcasted = downcast_to_lit_compatible(dtype.clone())?;
+
+    let len = values.len();
+    if len == 0 {
+        return Ok((Series::empty("literal", &dtype), None));
+    }
+    let mut errs: IndexMap<usize, String> = IndexMap::new();
+    let values = values.enumerate();
+
+    let field = Field::new("literal", downcasted.clone());
+
+    macro_rules! unwrap_inner {
+        ($expr:expr, $idx:expr, $variant:ident) => {
+            match $expr {
+                Ok(Literal::$variant(val)) => Some(val),
+                Ok(Literal::Null) => None,
+                Err(e) =>  {
+                    errs.insert($idx, e.to_string());
+                    None
+
+                },
+                Ok(other) => {
+                    let ty = other.get_type();
+                    errs.insert($idx, format!(
+                        "All literals must have the same data type or null. Found: {ty} vs {dtype}"
+                    ));
+                    None
+
+                }
+            }
+        };
+
+        ($expr:expr, $idx:expr, $literal:pat => $value:expr) => {
+            match $expr {
+                Ok($literal) => Some($value),
+                Ok(Literal::Null) => None,
+                Err(e) =>{
+                    errs.insert($idx, e.to_string());
+                    None
+                },
+                Ok(other) => {
+                    let ty = other.get_type();
+                    errs.insert($idx, format!(
+                        "All literals must have the same data type or null. Found: {ty} vs {dtype}"
+                    ));
+                    None
+                }
+            }
+        };
+    }
+
+    #[inline(always)]
+    fn from_mutable_primitive<T: NativeType, I: Iterator<Item = (usize, DaftResult<Literal>)>>(
+        values: I,
+        values_len: usize,
+        errs: &mut IndexMap<usize, String>,
+        field: Field,
+        f: impl Fn(&Literal) -> Option<T>,
+    ) -> DaftResult<Series> {
+        let mut arr = MutablePrimitiveArray::<T>::with_capacity(values_len);
+
+        for (i, value) in values {
+            let value = match value {
+                Ok(lit) => f(&lit),
+                Err(e) => {
+                    errs.insert(i, e.to_string());
+                    None
+                }
+            };
```

Comment:
This gives null instead of erroring if the literal does not match the expected data type, which is inconsistent with `unwrap_inner`.